dali/operators/reader/loader/video/video_file.cc

```diff
@@ -0,0 +1,232 @@
+// Copyright (c) 2021, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "dali/operators/reader/loader/video/video_file.h"
+#include "dali/core/error_handling.h"
+
+namespace dali {
+
+void VideoFileCPU::InitAvState() {
+  codec_ctx_ = avcodec_alloc_context3(codec_);
+  DALI_ENFORCE(codec_ctx_, "Could not create av codec context");
+
+  DALI_ENFORCE(
+    avcodec_parameters_to_context(codec_ctx_, codec_params_) >= 0,
+    "Could not fill the codec based on parameters");
+
+  DALI_ENFORCE(
+    avcodec_open2(codec_ctx_, codec_, nullptr) == 0,
+    "Could not initialize codec context");
+
+  frame_ = av_frame_alloc();
+  DALI_ENFORCE(frame_, "Could not allocate the av frame");
+
+  packet_ = av_packet_alloc();
+  DALI_ENFORCE(packet_, "Could not allocate av packet");
+}
+
+void VideoFileCPU::FindVideoStream() {
+  for (size_t i = 0; i < ctx_->nb_streams; ++i) {
+    auto stream = ctx_->streams[i];
+    codec_params_ = ctx_->streams[i]->codecpar;
+    codec_ = avcodec_find_decoder(codec_params_->codec_id);
+
+    if (codec_ == nullptr) {
+      continue;
+    }
+
+    if (codec_->type == AVMEDIA_TYPE_VIDEO) {
+        stream_id_ = i;
+        return;
+    }
+  }
+
+  DALI_FAIL(make_string("Could not find a valid video stream in file ", filename_));
+}
+
+VideoFileCPU::VideoFileCPU(std::string &filename) : 
+  filename_(filename) {
+  ctx_ = avformat_alloc_context();
+  DALI_ENFORCE(ctx_, "Could not create avformat context");
+
+  DALI_ENFORCE(
+    avformat_open_input(&ctx_, filename.c_str(), nullptr, nullptr) == 0,
```

Comment:
I would move this call from enforce and just check the result.

Reply:
Ad 1. Done
Ad 2. I'm aware of this situation. I want to handle it in a consecutive PR, so I can benchamrk the possible solutions properly. Basically, I want to thoroughly check which part takes how much time and what needs to be closed and what can remain allocated. The current solution generally allocates everything every time and I didn't want to port it as it is, since I suspect that it can be done better. But yes, this is the problem and it will be solved.